File: solutions/clojure/triangle/1/src/triangle.clj
(ns triangle)

(defn count-diff-sides [a b c]
  (let [[a b c :as sorted] (sort [a b c])]
    (when (and
           (every? pos? sorted)
           (< c (+ a b)))
      (count (set sorted)))))

(defn eq-sides [op n a b c]
  (let [cnt (count-diff-sides a b c)]
    (and (not (nil? cnt))
         (op n cnt))))

(def equilateral? (partial eq-sides = 1))

(def isosceles? (partial eq-sides >= 2))

(def scalene? (partial eq-sides = 3))

File: solutions/scala/triangle/1/src/main/scala/Triangle.scala

class Triangle(val a: Double, val b: Double, val c: Double) {
  def equilateral: Boolean = is_valid && sides.distinct.size == 1

  def isosceles: Boolean = is_valid && Set(1, 2)(sides.distinct.size)

  def scalene: Boolean = is_valid && sides.distinct.size == 3

  private def sides = List(a, b, c).sorted
  private def is_valid = sides.head + sides(1) > sides(2) && sides.forall(_ > 0)
}
